pkg/lexer/lexer_test.go

```diff
@@ -0,0 +1,252 @@
+			in:   "  x2 ",
+			want: &Token{Type: IDENT, Literal: "x2", Offset: 2, Line: 1, Col: 3},
+		}, "STRING": {
+			in:   `"xy"`,
+			want: &Token{Type: STRING_LIT, Literal: "xy", Offset: 0, Line: 1, Col: 1},
+		}, "STRING whitespace": {
+			in:   ` " x y "  `,
+			want: &Token{Type: STRING_LIT, Literal: " x y ", Offset: 1, Line: 1, Col: 2},
+		}, "STRING WITH COMMENT TOKEN": {
+			in:   `"x//y"`,
+			want: &Token{Type: STRING_LIT, Literal: "x//y", Offset: 0, Line: 1, Col: 1},
+		}, "STRING escaped quote": {
+			in:   `"xy\""`,
+			want: &Token{Type: STRING_LIT, Literal: `xy"`, Offset: 0, Line: 1, Col: 1},
+		}, "NUM": {
+			in:   " 1  ",
+			want: &Token{Type: NUM_LIT, Literal: "1", Offset: 1, Line: 1, Col: 2},
+		}, "NUM2": {
+			in:   "12 ",
+			want: &Token{Type: NUM_LIT, Literal: "12", Offset: 0, Line: 1, Col: 1},
+		}, "NUM3": {
+			in:   "12.3 ",
+			want: &Token{Type: NUM_LIT, Literal: "12.3", Offset: 0, Line: 1, Col: 1},
+		}, "NUM4": {
+			in:   "12.3.4 ",
+			want: &Token{Type: NUM_LIT, Literal: "12.3.4", Offset: 0, Line: 1, Col: 1},
+		}, "COMMENT": {
+			in:   "  // x2 ",
+			want: &Token{Type: COMMENT, Literal: "// x2 ", Offset: 2, Line: 1, Col: 3},
+		},
+	}
+
+	for name, tt := range tests {
+		tt := tt
+		t.Run(name, func(t *testing.T) {
+			l := New(tt.in)
+			got := l.Next()
+			assert.Equal(t, tt.want, got, "want %s got %s", tt.want.Type.String(), got.Type.String())
+
+			eofOffset := len(tt.in)
+			want := &Token{Type: EOF, Literal: "", Offset: eofOffset, Line: 1, Col: eofOffset + 1}
+			got = l.Next()
+			assert.Equal(t, want, got, "want %s got %s", tt.want.Type.String(), got.Type.String())
+		})
+	}
+}
+
+func TestNums(t *testing.T) {
+	in := `
+y := 1
+y = y * ( 3 + 7)
+`
+	wantTokens := []Token{
+		{Type: NL, Literal: "", Offset: 0, Line: 1, Col: 1},
+		{Type: IDENT, Literal: "y", Offset: 1, Line: 2, Col: 1},
+		{Type: DECLARE, Literal: "", Offset: 3, Line: 2, Col: 3},
+		{Type: NUM_LIT, Literal: "1", Offset: 6, Line: 2, Col: 6},
+		{Type: NL, Literal: "", Offset: 7, Line: 2, Col: 7},
+		{Type: IDENT, Literal: "y", Offset: 8, Line: 3, Col: 1},
+		{Type: ASSIGN, Literal: "", Offset: 10, Line: 3, Col: 3},
+		{Type: IDENT, Literal: "y", Offset: 12, Line: 3, Col: 5},
+		{Type: ASTERISK, Literal: "", Offset: 14, Line: 3, Col: 7},
+		{Type: LPAREN, Literal: "", Offset: 16, Line: 3, Col: 9},
+		{Type: NUM_LIT, Literal: "3", Offset: 18, Line: 3, Col: 11},
+		{Type: PLUS, Literal: "", Offset: 20, Line: 3, Col: 13},
+		{Type: NUM_LIT, Literal: "7", Offset: 22, Line: 3, Col: 15},
+		{Type: RPAREN, Literal: "", Offset: 23, Line: 3, Col: 16},
+		{Type: NL, Literal: "", Offset: 24, Line: 3, Col: 17},
+		{Type: EOF, Literal: "", Offset: len(in), Line: 4, Col: 1},
+	}
+	l := New(in)
+	for _, want := range wantTokens {
+		got := l.Next()
+		assert.Equal(t, &want, got, "want %s got %s", want.Type.String(), got.Type.String())
+	}
+}
+
+func TestStrings(t *testing.T) {
+	in := `
+s:string
+s = "abc"
+s = s[:1] // "bc"`
```

Comment:
It would be good to have a test that has a comment and a line after it to ensure comments are read correctly wrt the following line.